src/common/log.c

```diff
@@ -0,0 +1,273 @@
+ * custom log function - if not NULL (see rpma_log_open)
+ * all logs end up there regardless of logging threshold
+ */
+
+static logfunc *Custom_log_function = NULL;
+
+enum rpma_log_level Rpma_log_level = RPMA_LOG_NOTICE;
+enum rpma_log_level Rpma_log_print_level = RPMA_LOG_NOTICE;
+enum rpma_log_level Rpma_log_backtrace_level = RPMA_LOG_DISABLED;
+
+/* public librpma API */
+
+/*
+ * Initialize the logging module. Messages prior
+ * to this call will be dropped.
+ */
+void
+rpma_log_open(logfunc *custom_log_function)
+{
+	if (custom_log_function) {
+		Custom_log_function = custom_log_function;
+	} else {
+		openlog("rpma", LOG_PID, LOG_LOCAL7);
+	}
+}
+
+/*
+ * Close the currently active log. Messages after this call
+ * will be dropped.
+ */
+void
+rpma_log_close(void)
+{
+	if (!Custom_log_function) {
+		closelog();
+	} else {
+		Custom_log_function = NULL;
+	}
+}
+
+#ifdef RPMA_LOG_BACKTRACE_LVL
+static void
+log_unwind_stack(FILE *fp, enum rpma_log_level level)
+{
+	unw_error_t err;
+	unw_cursor_t cursor;
+	unw_context_t uc;
+	unw_word_t ip;
+	unw_word_t offp;
+	char f_name[64];
+	int frame;
+
+	if (level > Rpma_log_backtrace_level) {
+		return;
+	}
+
+	unw_getcontext(&uc);
+	unw_init_local(&cursor, &uc);
+	fprintf(fp, "*%s*: === BACKTRACE START ===\n", rpma_level_names[level]);
+
+	unw_step(&cursor);
+	for (frame = 1; unw_step(&cursor) > 0; frame++) {
+		unw_get_reg(&cursor, UNW_REG_IP, &ip);
+		err = unw_get_proc_name(&cursor, f_name, sizeof(f_name), &offp);
+		if (err || strcmp(f_name, "main") == 0) {
+			break;
+		}
+
+		fprintf(fp, "*%s*: %3d: %*s%s() at %#lx\n", \
+				rpma_level_names[level], frame, frame - 1, "", \
+				f_name, (unsigned long)ip);
+	}
+	fprintf(fp, "*%s*: === BACKTRACE END ===\n", rpma_level_names[level]);
+}
+
+#else
+#define log_unwind_stack(fp, lvl)
+#endif
+
+/*
+ * Provide actual time in a readable string
+ */
+static void
+get_timestamp_prefix(char *buf, size_t buf_size)
+{
+	struct tm *info;
+	char date[24];
+	struct timespec ts;
+	long usec;
+
+	clock_gettime(CLOCK_REALTIME, &ts);
+	info = localtime(&ts.tv_sec);
+	usec = ts.tv_nsec / 1000;
+
+	strftime(date, sizeof(date), "%Y-%m-%d %H:%M:%S", info);
+	snprintf(buf, buf_size, "[%s.%06ld] ", date, usec);
+}
+
+/*
+ * Write messages either to the syslog and to stderr
+ * or call user defined log function.
+ */
+void
+rpma_log(enum rpma_log_level level, const char *file, const int line,
+		const char *func, const char *format, ...)
+{
+	va_list ap;
+
+	va_start(ap, format);
+	rpma_vlog(level, file, line, func, format, ap);
+	va_end(ap);
+}
+
+/*
+ * Write messages either to the syslog and to stderr
+ * or call user defined log function.
+ */
+void
+rpma_vlog(enum rpma_log_level level, const char *file, const int line,
+		const char *func, const char *format, va_list ap)
+{
+	int severity = LOG_INFO;
+	char prefix[256];
+	char timestamp[45];
+	char message[MAX_TMPBUF];
+
+	if (Custom_log_function) {
+		Custom_log_function(level, file, line, func, format, ap);
```

Comment:
Make the default log function the rest of this code, and have it set as default. This will remove this condition. So all logs would go through the indirection, and there would simply be a default implementation.